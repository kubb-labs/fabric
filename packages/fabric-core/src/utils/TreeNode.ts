import type * as KubbFile from '../KubbFile.ts'

type BarrelData = {
  file?: KubbFile.File
  path: string
  name: string
}

export type Graph = {
  nodes: Array<{ id: string; label: string }>
  edges: Array<{ from: string; to: string }>
}

export class TreeNode<TData = unknown> {
  data: TData
  parent?: TreeNode<TData>
  children: Array<TreeNode<TData>> = []
  #cachedLeaves?: Array<TreeNode<TData>>

  constructor(data: TData, parent?: TreeNode<TData>) {
    this.data = data
    this.parent = parent
  }

  addChild(data: TData): TreeNode<TData> {
    const child = new TreeNode(data, this)
    this.children.push(child)
    this.#cachedLeaves = undefined // invalidate cached leaves
    return child
  }

  get leaves(): Array<TreeNode<TData>> {
    if (this.#cachedLeaves) return this.#cachedLeaves
    if (this.children.length === 0) return [this]

    const stack: Array<TreeNode<TData>> = [...this.children]
    const result: Array<TreeNode<TData>> = []

    for (const node of stack) {
      if (node.children.length) {
        for (const child of node.children) stack.push(child)
      } else {
        result.push(node)
      }
    }

    this.#cachedLeaves = result
    return result
  }

  forEach(callback: (node: TreeNode<TData>) => void): this {
    const stack: Array<TreeNode<TData>> = [this]

    for (const node of stack) {
      callback(node)
      if (node.children.length) {
        for (const child of node.children) stack.push(child)
      }
    }
    return this
  }

  findDeep(predicate: (node: TreeNode<TData>) => boolean): TreeNode<TData> | undefined {
    for (const leaf of this.leaves) {
      if (predicate(leaf)) return leaf
    }
    return undefined
  }

  static toGraph(root: TreeNode<BarrelData>): Graph {
    const nodes: Array<{ id: string; label: string }> = []
    const edges: Array<{ from: string; to: string }> = []

    root.forEach((node) => {
      nodes.push({
        id: node.data.path,
        label: node.data.name,
      })

      for (const child of node.children) {
        edges.push({
          from: node.data.path,
          to: child.data.path,
        })
      }
    })

    return { nodes, edges }
  }

  static fromFiles(files: Array<KubbFile.File>, rootFolder = ''): TreeNode<BarrelData> | null {
    const normalizePath = (p: string): string => p.replace(/\\/g, '/')
    const normalizedRoot = normalizePath(rootFolder)
    const rootPrefix = normalizedRoot.endsWith('/') ? normalizedRoot : `${normalizedRoot}/`

    const filteredFiles = files.filter((file) => {
      const filePath = normalizePath(file.path)

      return !filePath.endsWith('.json') && (!rootFolder || filePath.startsWith(rootPrefix))
    })

    if (filteredFiles.length === 0) {
      return null
    }

    const treeNode = new TreeNode<BarrelData>({
      name: rootFolder || '',
      path: rootFolder || '',
      file: undefined,
    })

    for (const file of filteredFiles) {
      const relPath = normalizePath(file.path).slice(rootPrefix.length)
      const parts = relPath.split('/')

      let current = treeNode
      let currentPath = rootFolder

      for (const [index, part] of parts.entries()) {
        const isLast = index === parts.length - 1
        currentPath += (currentPath.endsWith('/') ? '' : '/') + part

        let next: TreeNode<BarrelData> | undefined
        for (const child of current.children) {
          if ((child.data as BarrelData).name === part) {
            next = child
            break
          }
        }

        if (!next) {
          next = current.addChild({
            name: part,
            path: currentPath,
            file: isLast ? file : undefined,
          })
        }

        current = next
      }
    }

    return treeNode
  }
}
