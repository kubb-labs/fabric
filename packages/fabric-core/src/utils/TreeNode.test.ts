import path from 'node:path'
import type * as KubbFile from '../KubbFile.ts'
import { TreeNode } from './TreeNode.ts'

describe('TreeNode', () => {
  const files: KubbFile.File[] = [
    {
      path: 'src/test.ts',
      baseName: 'test.ts',
      sources: [],
      meta: {},
    },
    {
      path: 'src/sub/hello.ts',
      baseName: 'hello.ts',
      sources: [],
      meta: {},
    },
    {
      path: 'src/sub/world.ts',
      baseName: 'world.ts',
      sources: [],
      meta: {},
    },
  ]
  const tree = TreeNode.fromFiles(files, 'src/')
  const treeWindows = TreeNode.fromFiles(files, 'src\\')

  test('if schemas folder contains x files and y folders', () => {
    expect(tree).toBeDefined()
    expect(treeWindows).toBeDefined()

    expect(tree).toMatchSnapshot()
  })

  test('if leaves are rendered correctly', () => {
    expect(tree?.leaves.length).toBe(3)

    tree?.leaves.forEach((leave) => {
      if (leave.data.name === 'hello.ts') {
        expect(leave.data.path).toBe(path.join('src/sub', 'hello.ts'))
      }

      if (leave.data.name === 'hello.ts') {
        expect(leave.data.path).toBe(path.join('src/sub', 'hello.ts'))
      }

      if (leave.data.name === 'test.ts') {
        expect(leave.data.path).toBe(path.join('src/test.ts'))
      }
    })
  })
  test('if `find` is executed correctly', () => {
    const helloTS = tree?.leaves.find((leave) => leave.data.name === 'hello.ts')

    expect(tree?.findDeep).toBeDefined()
    expect(tree?.findDeep((item) => item.data === helloTS?.data)?.data.name).toEqual('hello.ts')
  })

  test('if `foreach` is executed correctly', () => {
    const items: TreeNode<{ name: string }>['data'][] = []

    tree?.forEach((treeNode) => {
      items.push(treeNode.data)
    })
    const names = items.map((item) => item.name)

    expect(items.length).toBe(5)
    expect(names).toMatchInlineSnapshot(`
      [
        "src/",
        "test.ts",
        "sub",
        "hello.ts",
        "world.ts",
      ]
    `)
  })

  test('if TreeNode.toGraph renders a graph correctly', () => {
    expect(tree).toBeDefined()

    expect(TreeNode.toGraph(tree!)).toMatchSnapshot()
  })
})
