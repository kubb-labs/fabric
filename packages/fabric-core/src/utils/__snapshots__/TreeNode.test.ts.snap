// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TreeNode > if TreeNode.toGraph renders a graph correctly 1`] = `
{
  "edges": [
    {
      "from": "src/",
      "to": "src/test.ts",
    },
    {
      "from": "src/",
      "to": "src/sub",
    },
    {
      "from": "src/sub",
      "to": "src/sub/hello.ts",
    },
    {
      "from": "src/sub",
      "to": "src/sub/world.ts",
    },
  ],
  "nodes": [
    {
      "id": "src/",
      "label": "src/",
    },
    {
      "id": "src/test.ts",
      "label": "test.ts",
    },
    {
      "id": "src/sub",
      "label": "sub",
    },
    {
      "id": "src/sub/hello.ts",
      "label": "hello.ts",
    },
    {
      "id": "src/sub/world.ts",
      "label": "world.ts",
    },
  ],
}
`;

exports[`TreeNode > if schemas folder contains x files and y folders 1`] = `
TreeNode {
  "children": [
    TreeNode {
      "children": [],
      "data": {
        "file": {
          "baseName": "test.ts",
          "meta": {},
          "path": "src/test.ts",
          "sources": [],
        },
        "name": "test.ts",
        "path": "src/test.ts",
      },
      "parent": [Circular],
    },
    TreeNode {
      "children": [
        TreeNode {
          "children": [],
          "data": {
            "file": {
              "baseName": "hello.ts",
              "meta": {},
              "path": "src/sub/hello.ts",
              "sources": [],
            },
            "name": "hello.ts",
            "path": "src/sub/hello.ts",
          },
          "parent": [Circular],
        },
        TreeNode {
          "children": [],
          "data": {
            "file": {
              "baseName": "world.ts",
              "meta": {},
              "path": "src/sub/world.ts",
              "sources": [],
            },
            "name": "world.ts",
            "path": "src/sub/world.ts",
          },
          "parent": [Circular],
        },
      ],
      "data": {
        "file": undefined,
        "name": "sub",
        "path": "src/sub",
      },
      "parent": [Circular],
    },
  ],
  "data": {
    "file": undefined,
    "name": "src/",
    "path": "src/",
  },
  "parent": undefined,
}
`;
